PT-1714 | feat: upgrade to HDS v3.5 & RHHC v1.0.0-alpha255 that uses HDS v3.5

karisal-anders opened the pull request:
## Description :sparkles:

---

**Tests** are **updated** in **another PR** as part of [PT-1725](https://helsinkisolutionoffice.atlassian.net/browse/PT-1725)

---

**NOTE**:
 - `REACT_APP_ORIGIN` has already been configured in all environments (review/test/staging/production)
 - The menus have been created in Headless CMS staging & production but not yet set to be in specific menu locations, which might make them not visible

### feat: upgrade to HDS v3.5 & RHHC v1.0.0-alpha255 that uses HDS v3.5

also:
 - make footer contents configurable from CMS
 - make header menu hierarchy configurable from CMS using normal CMS
   support i.e. drag & dropping menu items into a hierarchy in CMS
 - remove all feature flags as unnecessary
 - remove CIMODE language as unsupported by RHHC
 - remove unused SCSS files

refs PT-1714

### chore: optimize images

refs PT-1714

### feat: use HDS v3.5's hds-favicon-kit for favicons

refs PT-1714

### feat: use HDS v3.5 breakpoints if possible, otherwise custom breakpoints

changes to breakpoints:
 - direct mapping possible from old breakpoint to HDS v3.5 breakpoint:
   - xs -> s (no change in amount, just name, was and is 576px)
   - sm -> m (no change in amount, just name, was and is 768px)
 - custom breakpoints that were not mappable to HDS v3.5 breakpoints:
   - md -> l-plus (no change in amount, just name, was and is 1024px)
   - lg -> removed as unused (was 1200px)
   - xlg -> xxl (no change in amount, just name, was and is 1600px)
 - added missing breakpoints from HDS v3.5:
   - xs (320px)
   - l (992px)
   - xl (1248px)

resulting breakpoints:
 - xs = $breakpoint-xs = 320px
 - s = $breakpoint-s = 576px
 - m = $breakpoint-m = 768px
 - l = $breakpoint-l = 992px
 - l-plus = $breakpoint-l + 32px = 1024px
 - xl = $breakpoint-xl = 1248px
 - xxl = $breakpoint-xl + $breakpoint-xs + 32px = 1600px

refs PT-1714

### chore: replace custom icons with same or similar HDS v3.5 icons

replace custom icon with same HDS v3.5 icon:
 - IconCalendar -> IconCalendarPlus
 - IconClock -> IconClock
 - IconTicket -> IconTicket

replace custom icon with similar HDS v3.5 icon:
 - IconFood (crossed utensils) -> IconRestaurant (uncrossed utensils)

refs PT-1714

## Issues :bug:
### Closes :no_good_woman:
[PT-1714](https://helsinkisolutionoffice.atlassian.net/browse/PT-1714)

### Related :handshake:

## Testing :alembic:
### Automated tests :gear:️

**Tests** are **updated** in **another PR** as part of [PT-1725](https://helsinkisolutionoffice.atlassian.net/browse/PT-1725)

### Manual testing :construction_worker_man:

#### Manual testing in review environment

Can't be tested in review environment because review environment's tunnistamo doesn't support dynamic callback URLs and the review environment's URL is dynamic (i.e. based on the PR number).

#### Manual testing in local environment

Put this in your .env.development.local:
```
REACT_APP_OIDC_CLIENT_ID=kultus-provider-test
REACT_APP_ORIGIN=http://localhost:3000
REACT_APP_OIDC_AUTHORITY=https://tunnistamo.test.hel.ninja
REACT_APP_OIDC_CLIENT_ID=kultus-provider-test
REACT_APP_OIDC_SCOPE="openid profile https://api.hel.fi/auth/kultusapitest"
REACT_APP_API_URI=https://kultus.api.test.hel.ninja/graphql
REACT_APP_CMS_URI=https://kultus.hkih.stage.geniem.io/graphql
REACT_APP_LINKEDEVENTS_API_URI=https://api.hel.fi/linkedevents/v1
REACT_APP_SENTRY_DSN=
REACT_APP_ENVIRONMENT=local
REACT_APP_HEADLESS_CMS_ENABLED=true
```

And then "yarn" + "yarn build" + "yarn start".


## Screenshots :camera_flash:

## Additional notes :spiral_notepad:

[PT-1714]: https://helsinkisolutionoffice.atlassian.net/browse/PT-1714?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PT-1725]: https://helsinkisolutionoffice.atlassian.net/browse/PT-1725?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela commented on src/domain/app/layout/PageLayout.tsx:
```diff
@@ -20,7 +20,7 @@ import { getCmsPath } from '../routes/utils';
 import ProtectedPageWrapper from './ProtectedPageWrapper';
 
-export const MAIN_CONTENT_ID = 'main-content';
+export const MAIN_CONTENT_ID = 'main-content' as const;
```
It is already an unambiguous const, so there should be no need to reintroduce it as such.

<img width="821" alt="image" src="https://github.com/City-of-Helsinki/palvelutarjotin-admin/assets/389204/0b86ccb5-52c2-4d5c-8f7b-a2ff86b15669">
https://www.typescriptlang.org/play?#code/MYewdgzgLgBAsgQQJIDkD6BhA8igKgUTzSQBEYBeGAcgFsBDASzAFpQwoBTdqgbgCg+baPGTpseQrmJlKtRizaduMOhBhCoPIA

karisal-anders replied:
Learned something new today! Removed the unnecessary "as const" from string literals in the codebase.

nikomakela commented on src/utils/getLanguageCode.ts:
Does the `as const` do anything here? It does not change the compiled JS and it gives no advantage at least in VSCode. The way how you used the `as const satisfies Record` did change how it was interpreted as a readonly.

karisal-anders replied:
Hmm, yes, it didn't do anything, changed it to "as const satisfies" so now it's more specific type.

nikomakela commented:
There is something still wrong in the env-files, because I got this message when running locally
```
The client identifier (client_id) is missing or invalid.
```

karisal-anders commented:
Force pushed [changes](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/52921c1524c76bb147057354d31f09e8c0528771..5d77742577237d28a3eaea7babc912299c3ac47f):
- Removed remnants of `REACT_APP_HEADLESS_CMS_ENABLED` as unnecessary
- Removed remnants of `REACT_APP_LANGUAGE_CIMODE_VISIBLE` as unnecessary
- Fixed REACT_APP_CMS_URI in .env.development.local.example (`kultus.` prefix was missing)
- Removed "as const" from string literals as Typescript types them as specific string literals already without "as const"
- Made appLanguageToLanguageCode type more specific by using "as const satisfies"

karisal-anders commented:
> There is something still wrong in the env-files, because I got this message when running locally
> 
> ```
> The client identifier (client_id) is missing or invalid.
> ```

I'll make another PR about changing the .env files more extensively.

terovirtanen commented:
**kultus-admin-ui branch is deployed to platta: https://kultus-admin-ui-pr323.dev.hel.ninja** :rocket::rocket::rocket:

karisal-anders commented:
> > There is something still wrong in the env-files, because I got this message when running locally
> > ```
> > The client identifier (client_id) is missing or invalid.
> > ```
> 
> I'll make another PR about changing the .env files more extensively.

Created https://github.com/City-of-Helsinki/palvelutarjotin-admin/pull/324 for this
